### PR TITLE
[BugFix] Fix the get rowset by version return unexpect result

### LINKED ISSUE #12055: [Bug]get rowset by version return unexpect result
### Steps to reproduce the behavior (Required)
E.G there are rowsets in _rs_version_map: <1,1>,<2,2>,<3,4>, rowsets in _stale_rs_version_map: <3,3>,<4,4>, then _capture_consistent_rowsets_unlocked(<4,4>) maybe return rowset with version <3, 4>.

### Expected behavior (Required)
In `_capture_consistent_rowsets_unlocked`, first can't find <4,4> in _rs_version_map, then find <4,4> in _stale_rs_version_map.

### Reason

```
struct Version {
    int64_t first{0};
    int64_t second{0};

    Version(int64_t first_, int64_t second_) : first(first_), second(second_) {}
    Version() {}

    Version& operator=(const Version& version) {
        first = version.first;
        second = version.second;
        return *this;
    }

    friend std::ostream& operator<<(std::ostream& os, const Version& version);

    bool operator!=(const Version& rhs) const { return first != rhs.first || second != rhs.second; }

    bool operator==(const Version& rhs) const { return first == rhs.first && second == rhs.second; }

    bool contains(const Version& other) const { return first <= other.first && second >= other.second; }

    bool operator<(const Version& rhs) const { return second < rhs.second; }
};

# in tablet.h
class Tablet {
...
std::map<Version, RowsetSharedPtr> _rs_version_map;
...
}
```
The custom compare function of version is incorrect. Because equiv(a, b) is an expression equivalent to !comp(a, b) && !comp(b, a)  ( https://en.cppreference.com/w/cpp/named_req/Compare ).
So <3, 4> is equal to <4, 4> in `_rs_version_map `.

### StarRocks version (Required)
 - 2.1,2,2,2.3,2.4,2.5

### PR DESCRIPTION
**Pull request by luohaha**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #12055

## Problem Summary(Required) ：

E.G there are rowsets in _rs_version_map: <1,1>,<2,2>,<3,4>, rowsets in _stale_rs_version_map: <3,3>,<4,4>, then _capture_consistent_rowsets_unlocked(<4,4>) maybe return rowset with version <3, 4>. 
```
struct Version {
    int64_t first{0};
    int64_t second{0};
    ....
    bool operator<(const Version& rhs) const { return second < rhs.second; }
};

# in tablet.h
class Tablet {
...
std::map<Version, RowsetSharedPtr> _rs_version_map;
...
}
```
The custom compare function of version is incorrect. Because equiv(a, b) is an expression equivalent to !comp(a, b) && !comp(b, a)  ( https://en.cppreference.com/w/cpp/named_req/Compare ).
So <4, 4> is equal to <3, 4> in `_rs_version_map `.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: be/src/storage/tablet.cpp [hunk 1]
```diff
@@ -249,6 +249,26 @@ Status Tablet::add_rowset(const RowsetSharedPtr& rowset, bool need_persist) {
     return Status::OK();
 }
 
+// don't delete useless rowset, for ut
```
Comment (trueeyu) [inline be/src/storage/tablet.cpp:252]: Better not add func only for ut;

File: be/src/storage/tablet.cpp [hunk 2]
```diff
@@ -249,6 +249,26 @@ Status Tablet::add_rowset(const RowsetSharedPtr& rowset, bool need_persist) {
     return Status::OK();
 }
```
Comment (trueeyu) [inline be/src/storage/tablet.cpp:251]: if return wrong result, what will happend?

**Review by satanson — DISMISSED** [review]

File: be/src/storage/tablet.cpp (hunk 2 above)
Comment (luohaha) [reply be/src/storage/tablet.cpp:251]: I don't understand, newly add code didn't call `add_rowset()`, right?

**Review by meegoo — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: be/src/storage/olap_common.h [hunk 3]
```diff
@@ -376,7 +376,17 @@ struct Version {
 
     bool contains(const Version& other) const { return first <= other.first && second >= other.second; }
 
-    bool operator<(const Version& rhs) const { return second < rhs.second; }
+    bool operator<(const Version& rhs) const {
```
Comment (meegoo) [inline be/src/storage/olap_common.h:379]: could you add some comments about in which scene will exists overlap version in rs_version_map

**Review by meegoo — CHANGES_REQUESTED** [review]
could you add some comments about in which scene will exists overlap version in rs_version_map

**Review by meegoo — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: be/src/storage/tablet.cpp (hunk 1 above)
Comment (luohaha) [reply be/src/storage/tablet.cpp]: Done

**Review by satanson — APPROVED** [review]

**Comment by luohaha** [comment]
run starrocks_admit_test

**Review by trueeyu — APPROVED** [review]

File: be/src/storage/olap_common.h [hunk 4]
```diff
@@ -376,7 +376,17 @@ struct Version {
...
+        if (second < rhs.second) {
```
Comment (chaoyli) [inline be/src/storage/olap_common.h:380]:
```suggestion
        if (second == rhs.second) {
               return first < rhs.first;
        } else {
               return second < rhs.second;
        }
```

**Comment by luohaha** [comment]
run starrocks_admit_test

File: be/src/storage/olap_common.h (hunk 4 above)
Comment (luohaha) [reply be/src/storage/olap_common.h:380]: Maybe it is not that clear about what we want to do?

**Comment by meegoo** [comment]
@mergify backport branch-2.4 branch-2.3 branch-2.2

**Pull request merged**